File: 백준/Gold/21940. 가운데에서 만나기/가운데에서 만나기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[][] graph = new int[N+1][N+1];
        int result = 10000000;

        for (int i = 1; i < N+1; i++) {
            Arrays.fill(graph[i], 10000000);
            graph[i][i] = 0;
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            int T = Integer.parseInt(st.nextToken());

            graph[A][B] = T;
        }

        int K = Integer.parseInt(br.readLine());
        int[] friends = new int[K];
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < K; i++) {
            friends[i] = Integer.parseInt(st.nextToken());
        }

        for (int k = 1; k < N+1; k++) {
            for (int i = 1; i < N+1; i++) {
                for (int j = 1; j < N+1; j++) {
                    if (graph[i][j] > graph[i][k] + graph[k][j]) {
                        graph[i][j] = graph[i][k] + graph[k][j];
                    }
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 1; i < N+1; i++) {
            int curMax = 0;

            for (int j : friends) {
                curMax = Math.max(curMax, graph[i][j] + graph[j][i]);
            }

            if (curMax < result) {
                result = curMax;
                sb = new StringBuilder();
                sb.append(i).append(" ");
            } else if (curMax == result) {
                sb.append(i).append(" ");
            }
        }

        System.out.println(sb);
    }
}

File: 백준/Gold/21940. 가운데에서 만나기/가운데에서 만나기.py
# 가운데에서 만나기
import sys
input = sys.stdin.readline

N, M = map(int,input().split())
graph = [[float('inf')] * (N+1) for _ in range(N+1)]
result = float('inf')

for i in range(M):
    A, B, T = map(int,input().split())
    graph[A][B] = T

K = int(input())
friends = list(map(int,input().split()))

for i in range(1, N+1):
    graph[i][i] = 0

for k in range(1, N+1):
    for i in range(1, N+1):
        for j in range(1, N+1):
            if graph[i][j] > graph[i][k] + graph[k][j]:
                graph[i][j] = graph[i][k] + graph[k][j]

nodes = []

for i in range(1, N+1):
    curMax = 0

    for j in friends:
        curMax = max(curMax, graph[j][i] + graph[i][j])

    if curMax < result:
        result = curMax
        nodes = []
        nodes.append(i)
    elif curMax == result:
        nodes.append(i)
    
print(*nodes)
